pkg/cvo/metrics.go

```diff
@@ -83,6 +86,14 @@ version for 'cluster', or empty for 'initial'.
 			Name: "cluster_installer",
 			Help: "Reports info about the installation process and, if applicable, the install tool.",
 		}, []string{"type", "version", "invoker"}),
+		clusterVersionOperatorUpdateRetrievalTimestampSeconds: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "cluster_version_operator_update_retrieval_timestamp_seconds",
+			Help: "Reports when updates were last succesfully retrieved.",
+		}, []string{"name"}),
+		clusterVersionOperatorReasonFailure: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "cluster_version_operator_reason_failure",
+			Help: "Reports the reason when updates cannot be retrieved.",
+		}, []string{"reason"}),
```

Comment:
Surfacing [this previous thread][1] discussing the use of two metrics.  @simonpasquier in that comment:

> Also if the update fails for reason A then for reason B without updating the value of `cluster_version_operator_successful_update_retrieval` then the alert will reset and go back to pending because the reason label would have changed.

I'm not clear on why that would be the case, since with `cluster_version_operator_update_retrieval_timestamp_seconds` still over the threshold I'd expect the alert to continue to fire (but to switch to reporting the new reason).  But whatever, it seems to work the way we have it, and I can always float consolidation to a single metric in follow-up work if it bothers me and I can can get it to work that way ;).  This comment just gets a non-outdated discussion link where I can find it ;).

[1]: https://github.com/openshift/cluster-version-operator/pull/357#discussion_r429967805

Reply:
because the `reason` label attached to the alert would change, it would resolve the current alert and create a new one (just like timeseries, alerts are identified by their labels set).

Reply:
Ahhh, ok, that makes sense.  So the question is, do we want folks to have to renew a silence or whatever if the reason changes?  I guess I could go either way on that.  If we say "no, they should be able to silence once regardless of changing `reason`", then we should be able to get `reason` from the existing `cluster_operator_conditions` instead of defining this new `cluster_version_operator_reason_failure` metric.

Reply:
@wking @simonpasquier 

Use of cluster_operator_conditions makes for the following rather ugly expr and alert message constructs. Re-use of the metric cluster_operator_conditions is awkward. For example, after setting the channel and getting the VersionNotFound reason there are 10 cluster_operator_conditions metrics with name="version".

        message: Cluster version operator has not retrieved updates in {{ "{{ $value | humanizeDuration }}" }}. Failure reason {{ "{{ with $cluster_operator_conditions := \"cluster_operator_conditions\" | query}}{{range $value := .}}{{if and (eq (label \"name\" $value) \"version\") (eq (label \"condition\" $value) \"RetrievedUpdates\") (eq (label \"endpoint\" $value) \"metrics\") (eq (value $value) 0.0)}}{{label \"reason\" $value}} {{end}}{{end}}{{end}}" }}. {{ "{{ with $console_url := \"console_url\" | query }}{{ if ne (len (label \"url\" (first $console_url ) ) ) 0}} For more information refer to {{ label \"url\" (first $console_url ) }}/settings/cluster/.{{ end }}{{ end }}" }}
      expr: |
        (time()-cluster_version_operator_update_retrieval_timestamp_seconds) >= 120 and ignoring(condition, name, reason) cluster_operator_conditions{name="version", condition="RetrievedUpdates", endpoint="metrics", reason!="NoChannel"}

Reply:
You coud use `cluster_operator_conditions{name="version",condition="RetrievedUpdates"} == 0` as the initial query instead of `cluster_operator_conditions` and remove the inner `if` tests:

```
Failure reason {{ "{{ with $cluster_operator_conditions := \"cluster_operator_conditions{name=\\\"version\\\",condition=\\\"RetrievedUpdates\\\"} == 0\" | query}}{{label \"reason\" .}}{{end}}" }}. {{ "{{ with $console_url := \"console_url\" | query }}{{ if ne (len (label \"url\" (first $console_url ) ) ) 0}} For more information refer to {{ label \"url\" (first $console_url ) }}/settings/cluster/.{{ end }}{{ end }}" }}
```

Not sure about the escaping of the query though.

Reply:
> ... there are 10 cluster_operator_conditions metrics with name="version".

And `condition="RetrievedUpdates"`?  We should only set one of those at a time.  But maybe the PromQL service is giving us the last few?  Have you tried:

```
(time()-cluster_version_operator_update_retrieval_timestamp_seconds) >= 120 and ignoring(condition, name, reason) max(cluster_operator_conditions{name="version", condition="RetrievedUpdates", endpoint="metrics", reason!="NoChannel"})
```

Or some such?

Reply:
I cannot get any form of `cluster_operator_conditions{name="version",condition="RetrievedUpdates"} == 0 ` to work so I'm going with my original using the `if`.